src/modulus_convert.rs

```diff
@@ -0,0 +1,548 @@
+        p: u64,
+        batch_size: usize,
+    ) -> (Vec<ReplicatedSecretSharing>, Vec<u64>) {
+        // To send to helper 3
+        let mut values_to_share = Vec::with_capacity(batch_size);
+        let mut output_shares = Vec::with_capacity(batch_size);
+        for i in 0..(batch_size - 1) {
+            let alpha_2: u64 = sub_mod_p(
+                u64::from(r_binary_shares[i].sh1),
+                mult_mod_p(2, shares_of_r1r2[i].sh1, p),
+                p,
+            );
+            let s_1_2: u64 = rng_1_2.gen_range(0..p);
+            let s_2_3: u64 = rng_2_3.gen_range(0..p);
+            let d_2: u64 = sub_mod_p(alpha_2 * u64::from(r_binary_shares[i].sh2), s_1_2, p);
+            output_shares.push(ReplicatedSecretSharing {
+                sh1: sub_mod_p(alpha_2, mult_mod_p(2, s_1_2, p), p),
+                sh2: sub_mod_p(
+                    u64::from(r_binary_shares[i].sh2),
+                    mult_mod_p(2, add_mod_p(d_2, s_2_3, p), p),
+                    p,
+                ),
+            });
+            values_to_share.push(d_2);
+        }
+        (output_shares, values_to_share)
+    }
+
+    pub fn xor_r1_r2_r3_helper3<R: RngCore>(
+        mut rng_2_3: R,
+        r_binary_shares: &[ReplicatedBinarySecretSharing],
+        shares_of_r1r2: &[ReplicatedSecretSharing],
+        d2_values: &[u64],
+        p: u64,
+        batch_size: usize,
+    ) -> (Vec<ReplicatedSecretSharing>, Vec<u64>) {
+        // To send to helper 1
+        let mut d3_values = Vec::with_capacity(batch_size);
+        let mut output_shares = Vec::with_capacity(batch_size);
+        for i in 0..(batch_size - 1) {
+            let alpha_1 = sub_mod_p(
+                u64::from(r_binary_shares[i].sh2),
+                mult_mod_p(2, shares_of_r1r2[i].sh2, p),
+                p,
+            );
+            let s_2_3: u64 = rng_2_3.gen_range(0..p);
+            let r3: u64 = u64::from(r_binary_shares[i].sh1);
+            let d_3 = sub_mod_p(alpha_1 * r3, s_2_3, p);
+            d3_values.push(d_3);
+            output_shares.push(ReplicatedSecretSharing {
+                sh1: sub_mod_p(r3, mult_mod_p(2, add_mod_p(d2_values[i], s_2_3, p), p), p),
+                sh2: sub_mod_p(alpha_1, mult_mod_p(2, d_3, p), p),
+            });
+        }
+        (output_shares, d3_values)
+    }
+
+    // a and b should both be less than p
+    #[must_use]
+    pub fn add_mod_p(a: u64, b: u64, p: u64) -> u64 {
+        let offset: u64 = u64::MAX - p + 1;
+
+        let (sum, overlow) = a.overflowing_add(b);
+        if overlow {
+            return sum + offset;
+        } else if sum >= p {
+            return sum - p;
+        }
+        sum
+    }
+
+    // a and b should both be less than p
+    #[must_use]
+    pub fn sub_mod_p(a: u64, b: u64, p: u64) -> u64 {
+        let offset: u64 = u64::MAX - p + 1;
+
+        let (difference, overlow) = a.overflowing_sub(b);
+        if overlow {
+            return difference - offset;
```

Comment:
As above.  Rust people don't like using `return`.  (It takes some getting used to)